pkg/substitution/substitution.go

```diff
@@ -60,6 +60,22 @@ func ValidateVariableP(value, prefix string, vars sets.String) *apis.FieldError
 	return nil
 }
 
+func ValidateVariablePS(value, prefix string, suffix string, vars sets.String) *apis.FieldError {
+	if vs, present := extractVariablesFromString(value, prefix); present {
```

Comment:
I think there may be an exposure here.  The function is validating references starting with the given prefix, so in your case `$(tasks.`.   It is not considering the suffix (results) when gathering these.  There are other supported references that start with `$(tasks.` (e.g. result references) so I'm concerned it could extract those too and try to validate them which is not your intention, right?

Reply:
right, `status` is specified while calling this function in `suffix`.  After this extraction, suffix is trimmed before comparing against `vars`.

`ValidateVariablePS` is extracting prefix `tasks` and trimming suffix `status` and validates the left over value is one of the pipelineTasks (specified in `vars`).

you are right, this function would extract other variations (result reference) as well like you mentioned but result references are validated with `validateParamResults`. Any suggestion on how to better handle this?

Reply:
> Any suggestion on how to better handle this? If not, I would leave it as is since result references does go through validation separately.

I'm concerned that when finally tasks are allowed to use result references in the future, this code will break because it will reject those references.

The "quick and dirty" way to fix is would be to check if the reference at least has the suffix before trimming it.  If it doesn't have that suffix then it isn't the type of reference your are trying to check.  A more involved fix would be to have code specific to this type of reference that uses a regex for exact matching.  For example, resultref.go has code specific to parsing result references.  I understand though that you are trying to leverage what's already there.

Reply:
Thanks @GregDritschler, I have added an explicit check on the prefix and suffix before trimming it. 

I was actually looking at how task results are validated and to my surprise, the pipeline task names are not validated while validating the task results syntax i.e. `validateParamResults` validates the syntax $(tasks.mytask.results.total) but does not validate if `mytask` is actually defined as a pipeline task. Its not worrisome as such since Its captured in building `dag` function which fails with: 

```
validation failed: invalid value: couldn't add link between your-task and my-task: task your-task depends on my-task but my-task wasn't present in Pipeline: spec.tasks
```

Reply:
Thanks @pritidesai - I do not see the check in this function - perhaps you refer to https://github.com/tektoncd/pipeline/pull/3390/files#diff-78ff878240189b20e0ef82f2912c93c0b3c7b445865e04fa9b98aab851be6130R321 (L321)? However that check will only work for parameters in the "$(var)" format, but not if extra text or more variables are included.

Reply:
I think I reasonable fix here could be to have a variation of `extractVariablesFromString` that returns the stripped `suffix` so that it could be verified.
We should follow up on this PR to had such fix to make sure the code does not break when we add support for results like @GregDritschler mentioned.